llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -494,6 +499,148 @@ class DefaultStrategy : public PostPipelinerStrategy {
+  enum PriorityComponent {
+    None,
+    NodeNum,
+    Latest,
+    Critical,
+    Sibling,
+    LCDLatest
+  };
+  static const int Shift = 3;
+  static const int Size = 1 << Shift;
+  static const int Mask = Size - 1;
+
+private:
+  const std::string Name;
+  std::set<int> SuccSiblingScheduled;
+  std::function<bool(const SUnit &A, const SUnit &B)> Discriminators[Size] = {
+      [&](const SUnit &A, const SUnit &B) { return false; },
+      [&](const SUnit &A, const SUnit &B) { return A.NodeNum < B.NodeNum; },
+      [&](const SUnit &A, const SUnit &B) {
+        auto &IA = Info[A.NodeNum];
+        auto &IB = Info[B.NodeNum];
+        return IA.Latest < IB.Latest;
+      },
+      [&](const SUnit &A, const SUnit &B) {
+        auto &IA = Info[A.NodeNum];
+        auto &IB = Info[B.NodeNum];
+        return IA.NumPushedEarliest > IB.NumPushedEarliest;
+      },
+      [&](const SUnit &A, const SUnit &B) {
+        return SuccSiblingScheduled.count(A.NodeNum) >
+               SuccSiblingScheduled.count(B.NodeNum);
+      },
+      [&](const SUnit &A, const SUnit &B) {
+        auto &IA = Info[A.NodeNum];
+        auto &IB = Info[B.NodeNum];
+        return IA.LCDLatest < IB.LCDLatest;
+      },
+      [&](const SUnit &A, const SUnit &B) { return false; },
+      [&](const SUnit &A, const SUnit &B) { return false; },
+  };
+  std::vector<int> Priority;
+
+  bool better(const SUnit &A, const SUnit &B) override {
+    for (auto P : Priority) {
+      if (P == PriorityComponent::None) {
+        break;
+      }
+      if (Discriminators[P](A, B)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void selected(const SUnit &N) override {
+    // Promote the critical path
+    NodeInfo *Pushed = &Info[N.NodeNum];
+    while (Pushed->LastEarliestPusher >= 0) {
+      Pushed = &Info[Pushed->LastEarliestPusher];
+      Pushed->NumPushedEarliest++;
+    }
+
+    // Promote my siblings
+    for (auto &SDep : N.Succs) {
+      if (SDep.getKind() != SDep::Data) {
+        continue;
+      }
+      for (auto &PDep : SDep.getSUnit()->Preds) {
+        if (PDep.getKind() != SDep::Data) {
+          continue;
+        }
+        SuccSiblingScheduled.insert(PDep.getSUnit()->NodeNum);
+      }
+    }
+  }
+
+public:
+  std::string name() override { return Name; }
+  ConfigStrategy(ScheduleDAGInstrs &DAG, std::vector<NodeInfo> &Info,
+                 int Length, uint64_t Param)
+      : PostPipelinerStrategy(DAG, Info, Length),
+        Name("Config_" + std::to_string(Length) + "_" + std::to_string(Param)) {
+    while (Param) {
+      Priority.emplace_back(Param & Mask);
+      Param >>= Shift;
+    }
+  }
+};
+
+bool PostPipeliner::tryHeuristics() {
+  // The minimum length makes sure that every node has a range in which it
+  // can be scheduled
+  int MinLength = II;
+  for (int K = 0; K < NInstr; K++) {
+    auto &Node = Info[K];
+    while (Node.Earliest > Node.Latest + MinLength) {
+      MinLength += II;
+    }
+  }
```

Comment:
I find this a bit confusing to have in `tryHeuristics`. Should that be its own function like `computeMinLength` called within `schedule()` itself?

Reply:
As discussed, MinLength is local to the strategies. But will factor into separate method.